### DISCUSSION TITLE
Error in prisma.dapp.findMany() - Invalid Invocation and Conversion Error

### DISCUSSION THREAD

**Discussion by rishikesh-y** [prisma/prisma#22085] [Q&A] [answered]
### Question

I am encountering an issue with the `prisma.dapp.findMany()` function in my application. The error message suggests a problem with converting the "chains" field, which is expected to be a non-nullable type "Int," but is receiving an incompatible value of "null." The specific error code is 'P2032'.
As you can see below in model that chains field is a Int[], and it can also store the null value.
This is issue i am facing only on our production Database ( we are using supabase on prod ).
With staging Db ( aws postgres Db) and locally it is working fine.

### Code Snippet:
```typescript
const dapps = await prisma.dapp.findMany({
  where: {
    AND: [
      {
        isListed: true,
      },
      {
        category: {
          not: null,
        },
      },
    ],
  },
  include: {
    _count: {
      select: {
        dappInstalls: true,
        views: true,
        ratings: true,
        impressions: true,
      },
    },
  },
});
```
![Screenshot from 2023-11-23 11-32-54](https://github.com/prisma/prisma/assets/120367811/938a24fb-2849-4b6c-af0a-fa605059193f)

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
model Dapp {
  id                    Int                      @id @default(autoincrement())
  name                  String?
  description           String?
  external_url          String?
  image                 String?
  attributes            Json?
  appUrl                String?
  social                Json?
  downloadBaseUrls      Json?
  contracts             Json?
  images                Json?
  repoUrl               String?
  dappId                String                   @unique
  minAge                Int?
  isForMatureAudience   Boolean?
  isSelfModerated       Boolean?
  language              String[]
  version               String?
  versionCode           String?
  isListed              Boolean?
  listDate              DateTime?                @db.Timestamp(6)
  expiryDate            DateTime?                @db.Timestamp(6)
  availableOnPlatform   String[]
  geoRestrictions       Json?
  developer             Json?
  tags                  String[]
  chains                Int[]
  category              DappCategory?
  subCategory           String?
  packageId             String?
  walletApiVersion      String[]
  minted                Boolean?
  referredBy            Json?
  privacyPolicyUrl      String?
  support               Json?
  cdn                   Json?
  whitelistedForStores  String[]
  bannedForStores       String[]
  ownerAddress          String?
  tokenId               String?
  tokens                Json?
  created_at            DateTime                 @default(now())              @db.Timestamptz(3)
  updated_at            DateTime                 @updatedAt @default(now())   @db.Timestamptz(3)
  dappInstalls          DappInstall[] // This will create a one-to-many relationship
  views                 Views[]
  ratings               Ratings[]
  verification          Verification[]
  impressions           Impression[]
  document_with_weights Unsupported("tsvector")? @default(dbgenerated("''::tsvector"))

  @@index([dappId])
  @@index([name])
  @@index([document_with_weights], type: Gin)
  @@index([ownerAddress])
  @@index([isListed])
  @@index([availableOnPlatform])
  @@index([category])
  @@index([subCategory])
  @@index([geoRestrictions])
  @@index([chains])
  @@index([isForMatureAudience])
  @@index([minted])
  @@index([language])
  @@index([listDate])
  @@index([minAge])
  @@index([whitelistedForStores])
  @@index([bannedForStores])
}
```

```ts
// Add any relevant Prisma Client queries here
```

- OS:  Ubuntu
- Database:  PostgresSQL
- Node.js version:  18.12.0

- Run `prisma -v` to see your Prisma version and paste it
- 5.6.0

**Comment by ludralph** [comment]
Hi @rishikesh-y 👋 

Thank you for raising this question

> As you can see below in model that `chains` field is a Int[], and it can also store the null value.

The scalar list `Int[]` can only hold integer values. It can't store the `null` value like you have suggested. There is a [feature request](https://github.com/prisma/prisma/issues/15886) to allow storing null in list items.

Confirm whether the `chains` field in the production database contains null values. Review the data in that column to ensure it aligns with the schema definition. You may also want to implement error handling specifically for this scenario. You can add conditional checks in your code to handle situations where null values are encountered in the chains field.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by rishikesh-y** [reply]
Hey @ludralph  :wave: 
Thank's for replying
I have already checked my prod DB , in the DB there are some rows which have chains field like this 
![Screenshot from 2023-11-23 15-28-59](https://github.com/prisma/prisma/assets/120367811/6be212c1-01e9-4552-9b73-8d74cba05d90)
as you can see some rows doesn't have value of chains field but this is allowed, right?. Because chains field has a whole is Null instead of like this [ null, 137 ] . Also on our staging DB there are some rows which have same data as in the image, but on that it works.

**Reply by ludralph** [reply]
Hi @rishikesh-y 👋 

I have reproduced your issue.

The chains fields you defined in the Prisma schema, is of type `Int[]`. Within Prisma, this implies that `chains` as a field cannot be null or have null values. Within the underlying database, in this case Supabase, chains is actually nullable because that is how [Arrays](https://www.postgresql.org/docs/12/arrays.html#:~:text=To%20set%20an%20element%20of%20an%20array%20constant%20to%20NULL%2C%20write%20NULL%20for%20the%20element%20value.%20(Any%20upper%2D%20or%20lower%2Dcase%20variant%20of%20NULL%20will%20do.)%20If%20you%20want%20an%20actual%20string%20value%20%E2%80%9CNULL%E2%80%9D%2C%20you%20must%20put%20double%20quotes%20around%20it.) are defined in Postgres which Supabase sits on top of. I added null fields to chains directly to the the Database using a Database viewer tool Tableplus but could not do so using a Prisma `create` call. When I performed the findMany, I got the exact same error as you.

This error implies that Prisma attempts to convert the null value to an integer to match what is defined in the schema. One way to solve this would be to update all data in the `chains` field to the defined type using a script.

**Reply by rishikesh-y** [reply]
Hey @ludralph :wave: 
But why does it work in my local and staging databases? Additionally, the null values in the 'chains' field are not manually added by us; instead, they are default values provided by the database. When a row is inserted without a specific value for the 'chains' column, the database automatically assigns a null value to it as the default.

**Reply by ludralph** [reply, marked as answer]
HI @rishikesh-y 👋 

To clarify what I think is going on, I'm going to go through all the steps again.

1. When you create a new `Dapp` record, without specifying a value for `chains` field for example:
```js
const res = await prisma.dapp.create({
  data:{
    name: 'test',
    description: 'test',
  }
})
```
it will insert the record in the database and you would see `NULL` as can be seen in this screenshot
<img width="936" alt="image" src="https://github.com/prisma/prisma/assets/13855608/11f557b7-39e1-4d56-b10e-77d6a3dfec34">
2. When you run `findMany` on the database, it will return the record as shown: it coarsces the `NULL` to an empty array `[]` which is what the [issue](https://github.com/prisma/prisma/issues/13880) hopes to address.  I think your staging and local databases are behaving in this manner which explains why there is no error there
```js
prisma:query SELECT "public"."Dapp"."id", "public"."Dapp"."name", "public"."Dapp"."description", "public"."Dapp"."chains" FROM "public"."Dapp" WHERE 1=1 OFFSET $1
{ id: 6, name: 'test', description: 'test', chains: [] }
[
  { id: 1, name: 'test', description: 'test', chains: [] },
  { id: 2, name: 'test', description: 'test', chains: [] },
  { id: 3, name: 'text', description: 'text', chains: [] },
  { id: 4, name: 'text', description: 'text', chains: [] },
  { id: 6, name: 'test', description: 'test', chains: [] }
]
```
3. At this point, I manually add an array directly to the DB which has null in the second index of the array [1,null]. Look at the 7th record in the image below.
<img width="1076" alt="image" src="https://github.com/prisma/prisma/assets/13855608/15e55a35-aeab-48a2-b81b-1a3ac5cc0d98">

Now when I run the call to `findMany` on the `Dapps` record, the error is thrown
```
 83 const res1 = await prisma.dapp.findMany(
Error converting field "chains" of expected non-nullable type "Int", found incompatible value of "null".
    at ni.handleRequestError (/Users/raphaeletim/Documents/prisma-graphql/node_modules/@prisma/client/runtime/library.js:124:6817)
    at ni.handleAndLogRequestError (/Users/raphaeletim/Documents/prisma-graphql/node_modules/@prisma/client/runtime/library.js:124:6206)
    at ni.request (/Users/raphaeletim/Documents/prisma-graphql/node_modules/@prisma/client/runtime/library.js:124:5926)
    at async l (/Users/raphaeletim/Documents/prisma-graphql/node_modules/@prisma/client/runtime/library.js:129:10023)
    at async main (/Users/raphaeletim/Documents/prisma-graphql/src/script.js:83:14) {
  code: 'P2032',
  clientVersion: '5.6.0',
  meta: { field: 'chains', expected_type: 'Int', found: 'null' }
}
```
My guess is there are records in Prod DB that have null as a value in an index in the array that is why that error is showing up.

**Reply by rishikesh-y** [reply]
Thanks, @ludralph, There are some rows that have chains field like this [null]. I updated those rows, and now the issue is resolved

**Reply by Karthik73965** [reply]
Got the same issue and solved thanks @rishikesh-y for creating a issue
